File: rust/src/bot/cache.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use solana_sdk::pubkey::Pubkey;
use std::collections::HashMap;
use std::time::Instant;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Cache {
    pub start_time: DateTime<Utc>,
    pub queue: HashMap<u64, i8>,
    pub queue_throttle: u32,
    pub side_buy: bool,
    pub iteration: u64,
    pub wallet_pubkey: String,
    pub wallet_pubkey_full: String,
    
    pub iteration_per_minute: IterationStats,
    pub initial_balance: Balance,
    pub current_balance: Balance,
    pub current_profit: Profit,
    pub last_balance: Balance,
    pub profit: Profit,
    pub max_profit_spotted: MaxProfit,
    pub trade_counter: TradeCounter,
    pub ui: UiConfig,
    pub chart: ChartData,
    pub hotkeys: Hotkeys,
    
    pub trading_enabled: bool,
    pub wrap_unwrap_sol: bool,
    pub swapping_right_now: bool,
    pub fetching_results: bool,
    pub fetching_results_start: Instant,
    pub trade_history: Vec<TradeHistory>,
    pub performance_of_tx_start: Instant,
    pub available_routes: AvailableRoutes,
    pub is_setup_done: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IterationStats {
    pub start: Instant,
    pub value: f64,
    pub counter: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Balance {
    pub token_a: u64,
    pub token_b: u64,
}

// Add other required structs...

impl Default for Cache {
    fn default() -> Self {
        Self {
            start_time: Utc::now(),
            queue: HashMap::new(),
            queue_throttle: 1,
            side_buy: true,
            iteration: 0,
            // Initialize other fields...
        }
    }
}

impl Cache {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn update_iteration_per_minute(&mut self) {
        // Implementation
    }

    pub fn record_trade(&mut self, trade: TradeHistory) {
        self.trade_history.push(trade);
    }
} 

File: copy/src/bot/cache.js
// global cache
const cache = {
	startTime: new Date(),
	queue: {},
	queueThrottle: 1,
	sideBuy: true,
	iteration: 0,
	walletpubkey: '',
	walletpubkeyfull: '',
	iterationPerMinute: {
		start: performance.now(),
		value: 0,
		counter: 0,
	},
	initialBalance: {
		tokenA: 0,
		tokenB: 0,
	},

	currentBalance: {
		tokenA: 0,
		tokenB: 0,
	},
	currentProfit: {
		tokenA: 0,
		tokenB: 0,
	},
	lastBalance: {
		tokenA: 0,
		tokenB: 0,
	},
	profit: {
		tokenA: 0,
		tokenB: 0,
	},
	maxProfitSpotted: {
		buy: 0,
		sell: 0,
	},
	tradeCounter: {
		buy: { success: 0, fail: 0 },
		sell: { success: 0, fail: 0 },
		failedbalancecheck: 0,
		errorcount: 0,
	},
	ui: {
		defaultColor: process.env.UI_COLOR ?? "cyan",
		showPerformanceOfRouteCompChart: false,
		showProfitChart: false,
		showTradeHistory: false,
		hideRpc: false,
		showHelp: false,
		allowClear: true,
	},
	chart: {
		spottedMax: {
			buy: new Array(120).fill(0),
			sell: new Array(120).fill(0),
		},
		performanceOfRouteComp: new Array(120).fill(0),
	},
	hotkeys: {
		e: false,
		r: false,
	},
	tradingEnabled:
		process.env.TRADING_ENABLED === undefined
			? true
			: process.env.TRADING_ENABLED === "true",
	wrapUnwrapSOL:
		process.env.WRAP_UNWRAP_SOL === undefined
			? true
			: process.env.WRAP_UNWRAP_SOL === "true",
	swappingRightNow: false,
	fetchingResultsFromSolscan: false,
	fetchingResultsFromSolscanStart: 0,
	tradeHistory: [],
	performanceOfTxStart: 0,
	availableRoutes: {
		buy: 0,
		sell: 0,
	},
	isSetupDone: false,
	
	// Token rotation state
	tokenRotationList: [],
	currentRotationToken: null,
	currentRotationIndex: 0,
	
	config: {
		// Default configuration that will be replaced during setup
		rpc: [process.env.DEFAULT_RPC || ""],
		minInterval: parseInt(process.env.MIN_INTERVAL_MS) || 100,
		slippage: parseInt(process.env.MAX_SLIPPAGE_PERCENT * 100) || 100,
		priority: parseInt(process.env.PRIORITY) || 100,
		minPercProfit: parseFloat(process.env.MIN_PROFIT_THRESHOLD) || 0.5,
		adaptiveSlippage: process.env.ADAPTIVE_SLIPPAGE === "true" ? 1 : 0,
		tradingStrategy: "arbitrage",
		tradeSize: {
			value: parseFloat(process.env.TRADE_SIZE_SOL) || 1.0,
			strategy: process.env.TRADE_SIZE_STRATEGY || "fixed",
		},
		ui: {
			defaultColor: process.env.UI_COLOR || "cyan",
		},
		storeFailedTxInHistory: true,
	}
};

module.exports = cache;
